docs/tutorial.md

```diff
@@ -0,0 +1,656 @@
+
+Django provides tools to fix this. In this case, we'd probably use the `select_related` queryset method in the view, like this:
+
+```python
+def book_list(request):
+    queryset = Book.objects.select_related("publisher")
+    return render("book_list.html", {"books": queryset})
+```
+
+This tells the ORM to issue a `JOIN` to fetch all the data in one query:
+
+```sql
+SELECT "books_book"."id",
+       "books_book"."title",
+       "books_book"."publisher_id",
+       "books_book"."publication_date",
+       "books_publisher"."id",
+       "books_publisher"."name",
+       "books_publisher"."address"
+FROM "books_book"
+LEFT OUTER JOIN "books_publisher"
+ON ("books_book"."publisher_id" = "books_publisher"."id")
+```
+
+The problem with this is that the view code and the template code are now _coupled_ in a very unpredictable and hard-to-reason-about way. A small, innocuous-looking change to the code in either the view or template can drastically impact the performance. This coupling problem only gets worse as the complexity of the application increases.
+
+!!! note
+    YAGNI, "You Aren't Gonna Need It", it's a [well-understood mantra](https://martinfowler.com/bliki/Yagni.html) in software development. It means that you should only make changes to your software (including adding abstraction and generalising code structures) when you are sure that you absolutely need them, and not before. Related to this is the famous quote from Donald Knuth: "premature optimization is the root of all evil". This suggests that usually performance concerns fall under YAGNI: you shouldn't spend time making code fast until its slowness becomes a real problem for users.
+
+    As a counterpoint to this advice, Simon Willison coined the phrase [PAGNI](https://simonwillison.net/2021/Jul/1/pagnis/), "Probably Are Gonna Need It". PAGNI applies in situations "when the cost of adding something later is so dramatically expensive compared with the cost of adding it early on that it’s worth taking the risk [...] when you know from experience that an initial investment will pay off many times over."
+
+    `django-readers` (and its sister project `django-zen-queries`) were built partly as a result of a strong belief (gained through many years of experience) that understanding and controlling your application's database querying behaviour firmly falls into the PAGNI category. This is particularly true of applications that you know are going to be complex: any code abstraction patterns you decide to follow in order to manage the complexity _must_ take into account query patterns or they are highly likely to cause terrible performance problems. This is the heart of the [object-relational impedance mismatch](https://en.wikipedia.org/wiki/Object%E2%80%93relational_impedance_mismatch).
```

Comment:
This bit is lovely and quite an important part of the motivation behind `django-readers`

I feel like it needs a more prominent place, perhaps in the `Explanation`.

Though it does read nicely here as well.

Reply:
@j4mie I strongly agree with @nvlaarhoven about the following

> I feel like it needs a more prominent place, perhaps in the Explanation.

Which prominent place is up to you. BUt yes, this should be somewhere more prominent.

Reply:
Thanks for mentioning `zen-queries`, @nvlaarhoven . I had starred the repo but completely forgot about it until you pointed this out again. Thank you. 🙇🏻‍♂️

Reply:
I'll move this to the Explanation page.